File: src/edu/psu/ist/plato/kaiming/Procedure.java
package edu.psu.ist.plato.kaiming;

import java.util.Collection;
import java.util.List;

public abstract class Procedure {
    /**
     * A procedure is required to have a CFG, even if it is only a trivial one.
     */
    protected CFG mCfg;
    protected Label mLabel;

    public Procedure(Label l, List<? extends Entry> entries) {
        mLabel = l;
        mCfg = buildCFGInternal(entries);
    }

    public void setEntries(List<? extends Entry> entries) {
        mCfg = buildCFGInternal(entries);
    }

    public List<? extends Entry> getEntries() {
        return mCfg.getEntries();
    }

    public abstract CFG buildCFGInternal(List<? extends Entry> entries);

    public CFG getCFG() {
        return mCfg;
    }

    public abstract String getName();

    public abstract Label deriveSubLabel(BasicBlock bb);

    protected CFG createCFGObject(Collection<BasicBlock> bbs,
            BasicBlock entry) {
        return new CFG(bbs, entry);
    }
}

File: src/main/scala/edu/psu/ist/plato/kaiming/Procedure.scala
package edu.psu.ist.plato.kaiming

import ir.{Context, IRCfg}

abstract class Procedure[A <: Arch] {

  def label: Label
  def cfg: Cfg[A, _ <: BBlock[A]]
  def name = label.name
  def entries: Vector[Entry[A]]
  def index = cfg.entryBlock.index
  def deriveLabelForIndex(index: Long): Label = Label("_sub_" + index.toHexString)

}

abstract class MachProcedure[A <: Arch](val entries: Vector[MachEntry[A]],
    trimIsolated: Boolean = false) extends Procedure[A] {
  
  def mach: Machine[A]
  override val cfg = MachCfg(this)
  
}